akka-typed/src/main/scala/akka/typed/internal/DispatchersImpl.scala

```diff
@@ -5,32 +5,223 @@ package akka.typed
+  }
+
+  private def idConfig(id: String): Config = {
+    import scala.collection.JavaConverters._
+    ConfigFactory.parseMap(Map("id" → id).asJava)
+  }
+
+  /**
+   * INTERNAL API
+   *
+   * Creates a MessageDispatcherConfigurator from a Config.
+   *
+   * The Config must also contain a `id` property, which is the identifier of the dispatcher.
+   *
+   * Throws: IllegalArgumentException if the value of "type" is not valid
+   *         IllegalArgumentException if it cannot create the MessageDispatcherConfigurator
+   */
+  private def configuratorFrom(cfg: Config): MessageDispatcherConfigurator = {
+    if (!cfg.hasPath("id")) throw new ConfigurationException("Missing dispatcher 'id' property in config: " + cfg.root.render)
+
+    cfg.getString("type") match {
+      case "Dispatcher"       ⇒ new DispatcherConfigurator(cfg, prerequisites)
+      case "PinnedDispatcher" ⇒ new PinnedDispatcherConfigurator(cfg, prerequisites)
+      case fqn ⇒
+        val args = List(classOf[Config] → cfg)
+        prerequisites.dynamicAccess.createInstanceFor[DispatcherConfigurator](fqn, args).recover({
+          case exception ⇒
+            throw new ConfigurationException(
+              ("Cannot instantiate DispatcherConfigurator type [%s], defined in [%s], " +
+                "make sure it has constructor with [com.typesafe.config.Config] and " +
+                "[akka.dispatch.DispatcherPrerequisites] parameters")
+                .format(fqn, cfg.getString("id")), exception)
+        }).get
+    }
+  }
+}
+
+/**
+ * Base class to be used for hooking in new dispatchers into Dispatchers.
+ */
+abstract class MessageDispatcherConfigurator(_config: Config, val prerequisites: d.DispatcherPrerequisites) {
+
+  val config: Config = new d.CachingConfig(_config)
+
+  /**
+   * Returns an instance of MessageDispatcher given the configuration.
+   * Depending on the needs the implementation may return a new instance for
+   * each invocation or return the same instance every time.
+   */
+  def dispatcher(): ExecutionContextExecutor
+
+  def configureExecutor(): d.ExecutorServiceConfigurator = {
+    def configurator(executor: String): d.ExecutorServiceConfigurator = executor match {
+      case null | "" | "fork-join-executor" ⇒ new d.ForkJoinExecutorConfigurator(config.getConfig("fork-join-executor"), prerequisites)
+      case "thread-pool-executor"           ⇒ new d.ThreadPoolExecutorConfigurator(config.getConfig("thread-pool-executor"), prerequisites)
+      case fqcn ⇒
+        val args = List(
+          classOf[Config] → config,
+          classOf[d.DispatcherPrerequisites] → prerequisites)
+        prerequisites.dynamicAccess.createInstanceFor[d.ExecutorServiceConfigurator](fqcn, args).recover({
+          case exception ⇒ throw new IllegalArgumentException(
+            ("""Cannot instantiate ExecutorServiceConfigurator ("executor = [%s]"), defined in [%s],
+                make sure it has an accessible constructor with a [%s,%s] signature""")
```

Comment:
This code has been copied from akka-actor, shall I fix it there as well? (we might have multiple leftover cases in there, perhaps better as a separate ticket)